### PR TITLE
[TRTLLM-8031][feat] Add chunked return_generation_logits logic

### PR DESCRIPTION
**Pull request by yibinl-nvidia**
## Summary by CodeRabbit

* **New Features**
  * Added optional chunked transfer for logits to reduce host memory usage during long generations.
  * Configurable enable/disable and chunk size via request options; applies to both context and generation logits, with streaming and non‑streaming support.
  * Added a post‑processing step to finalize any remaining chunked logits so results are complete and accessible.

* **Tests**
  * Introduced comprehensive unit, integration, and performance tests for chunked logits.

## Description
Context logits handling is unchanged. This MR only affects generation logits storage in LlmRequest.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by tijyojwad — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 1]
```diff
@@ -193,6 +275,13 @@ def append_mm_embeddings(self, mm_embeddings: torch.Tensor):
         self._mm_embeddings = SharedTensorContainer.from_tensor(
             mm_embeddings).dump_to_dict()
 
+    def post_processing_transfer(self):
```
Comment (venkywonka) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:278]: `post_processing_transfer` maybe is too generic? like it doesn't tell anything about the scope of what it transfers... maybe calling it `transfer_remaining_device_logits` or something like that? just a soft suggestion

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 2]
```diff
@@ -108,6 +140,44 @@ def get(self, all_logits: bool) -> torch.Tensor | None:
...
+        # Update position and clear fragments
+        self._logits_indices.append((self._current_position, end_pos))
+        self._current_position = end_pos
+        self._device_fragments.clear()
+
+    def finalize_transfer(self):
+        """Force transfer of any remaining fragments to host (for chunked mode)"""
+        if self.use_chunked_logits and hasattr(
+                self, '_device_fragments') and self._device_fragments:
+            self._transfer_chunk_to_host()
```
Comment (venkywonka) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:179]: technically `finalize_transfer` can be absorbed inside `self._transfer_chunk_to_host()` and that can be called directly in `PyResult`  right? might simplify things a bit?

File: tests/unittest/_torch/executor/test_chunked_logits.py [hunk 3]
```diff
@@ -0,0 +1,943 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/executor/test_chunked_logits.py]: fixed

File: tests/unittest/_torch/executor/test_chunked_logits.py [hunk 4]
```diff
@@ -0,0 +1,943 @@
...
+        # Should have fragment added
+        assert len(storage._device_fragments) == 1
+        assert storage._current_position == 0
+
+    @pytest.mark.unit
+    def test_append_chunked_mode_streaming(self, sample_logits):
+        """Test append behavior in chunked streaming mode"""
+        storage = LogitsStorage(seq_length=10,
+                                use_chunked_logits=True,
+                                streaming=True)
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/executor/test_chunked_logits.py]: fixed

File: tests/unittest/_torch/executor/test_chunked_logits.py [hunk 5]
```diff
@@ -0,0 +1,943 @@
...
+        executor_request.mrope_config = None
+        executor_request.multimodal_embedding = None
+        executor_request.lora_config = None
+        executor_request.guided_decoding_params = None
+        executor_request.context_phase_params = None
+        executor_request.cache_salt_id = None
+        executor_request.client_id = None
+        executor_request.request_type = tllm_executor.RequestType.REQUEST_TYPE_CONTEXT_AND_GENERATION
+        executor_request.use_chunked_logits = True
+        executor_request.logits_chunk_size = 4
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/executor/test_chunked_logits.py]: not relavant anymore. Code removed.

File: tests/unittest/_torch/executor/test_chunked_logits.py [hunk 6]
```diff
@@ -0,0 +1,943 @@
...
+
+        # Add second fragment - should trigger transfer
+        storage.append(sample_logits)
+        assert len(storage._device_fragments) == 0
+        assert storage._current_position == 2
+
+    @pytest.mark.unit
+    def test_get_all_logits(self, sample_logits):
+        """Test get method with all_logits=True"""
+        storage = LogitsStorage(seq_length=10, use_chunked_logits=False)
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/executor/test_chunked_logits.py]: I decided to remove `test_get_all_logits` because the logic for `get()` isn't changed. Good advice though!

File: tests/unittest/_torch/executor/test_chunked_logits.py [hunk 7]
```diff
@@ -0,0 +1,943 @@
...
+        storage.append(sample_logits)
+
+        result = storage.get(all_logits=True)
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/executor/test_chunked_logits.py]: same as above, `get()` function logic is unchanged and can be covered by integration test.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 1 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py]: Thanks for the suggestion. Modified function name.

**Review by venkywonka — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 8]
```diff
@@ -108,6 +140,43 @@ def get(self, all_logits: bool) -> torch.Tensor | None:
...
+
+    def _transfer_chunk_to_host(self):
+        """Transfer accumulated fragments to host"""
+        if not self._device_fragments:
+            return
+
+        # Allocate host storage if needed
+        assert self._storage is not None, "Storage should be initialized"
+        # if self._storage is None:
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:159]: Maybe forgot to delete?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 9]
```diff
@@ -165,13 +234,25 @@ def __init__(self,
                  return_log_probs: bool = False,
                  return_context_logits: bool = False,
                  return_generation_logits: bool = False,
-                 exclude_last_generation_logits: bool = False):
+                 exclude_last_generation_logits: bool = False,
+                 use_chunked_generation_logits: bool = True,
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:238]: Can I ask why the default value is True for this newly introduced one?

**Review by yechank-nvidia — APPROVED** [review]
LGTM. minor comments

**Review by schetlur-nv — APPROVED** [review]
Rubber stamp given other approvals

**Review by dcampora — APPROVED** [review]
LGTM

**Review by tijyojwad — APPROVED** [review]
